Modify job types to respect additional namenode delegation token configs

xkrogen opened the pull request:
Modify the Hadoop(Spark,Hive,Pig,Java)Job types to respect the MapReduce- and Spark-level configurations for additional namenodes from which to fetch delegation tokens, in addition to the Azkaban-level other_namenodes configuration. MapReduce and Spark both have configurations for specifying namenodes from which additional delegation tokens should be fetched before submitting a job (`mapreduce.job.hdfs-servers` and `spark.yarn.access.namenodes`, respectively) but previously Azkaban ignored these, using it's own `other_namenodes` property. Azkaban should additionally respect these framework-level configurations.

HappyRay commented on plugins/jobtype/src/azkaban/jobtype/HadoopJobUtils.java:
```diff
@@ -85,6 +85,12 @@
+  public static final String MAPREDUCE_JOB_OTHER_NAMENODES = "mapreduce.job.hdfs-servers";
+
+  // Azkaban property for listing additional namenodes for delegation tokens
+  private static final String OTHER_NAMENODES_PROPERTY = "other_namenodes";
```
It would be good to share the definition with Azkaban core.

xkrogen replied:
Right now it's private in Azkaban core. If you want we can release a new version of Azkaban core which exports this property.

HappyRay commented on plugins/jobtype/test/azkaban/jobtype/TestHadoopSparkJob.java:
```diff
@@ -142,4 +142,31 @@ public void testSparkPatternLibConf() {
     Assert.assertEquals(sparkHomeConfPath[1], sparkConfPath);
   }
+
+  @Test
```
Nice. Thanks for adding tests!

HappyRay commented on plugins/jobtype/src/azkaban/jobtype/HadoopPigJob.java:
```diff
@@ -112,6 +113,11 @@ public void run() throws Exception {
       Props props = new Props();
       props.putAll(getJobProps());
       props.putAll(getSysProps());
+      String additionalNamenodes =
```
This code is the same in multiple classes. It would be great to reduce the code duplication.

xkrogen replied:
Nearly all of the code is duplicated between Hadoop(Pig,Hive,Java)Job. They really should inherit from a common subclass. I can do that as part of this PR if you'd like.

HappyRay commented:
In addition to the unit tests, what other tests do you plan to run?

suvodeep-pyne commented:
@xkrogen 
AFAIK. The spark code is moving outside `azkaban-plugins`. I think this might break if/when that happens?

xkrogen commented:
@HappyRay About other testing, I wanted to ask you about that. Do we have any good way to integration test these? I'm not familiar with the Azkaban develop/test process.

@suvodeep-pyne Where is it moving? Can we replicate this same logic there?

suvodeep-pyne commented:
@xkrogen 
It is moving to an internal repo as far as I am aware. @Victsm can provide some context.

Victsm commented:
@xkrogen 

We are moving the code outside of this repo in order to enable hot fix deployment for Spark job type plugin.
Azkaban team cannot hot deploy Spark job type plugin if the plugin is inside this repo.
We need to be able to generate Spark job type plugin JAR and publish it to artifactory in order to enable hot deployment which cuts down the delay for deploying changes to Spark job type.

xkrogen commented:
@Victsm Okay, sounds good. We need this fix to be deployed by next week. I'm assuming the change will not happen by then? In any case we need this fix in the SparkJob plugin wherever it lives.

xkrogen commented:
Updated the code to fix an issue with the previous code where it was relying on Spark being on the classpath (which it is not). Also refactored a bit per @HappyRay 's comments. I tested this on a live Azkaban cluster and confirmed that both Hadoop-type and Spark-type jobs respected this configuration successfully.

Victsm commented on plugins/jobtype/src/azkaban/jobtype/HadoopSparkJob.java:
```diff
@@ -175,6 +188,35 @@ public HadoopSparkJob(String jobid, Props sysProps, Props jobProps, Logger log)
+   * @param props Props to add additional namenodes to.
+   * @see HadoopJobUtils#addAdditionalNamenodesToProps(Props, String)
+   */
+  void addAdditionalNamenodesFromConf(Props props) {
```
Can we use the method getSparkProperties inside HadoopSecureSparkWrapper to get this config param we need?
We can change the scope of that method to make it accessible here.
No need to duplicate the logic for fetching configs inside spark-defaults.conf.

Also, once the new Azkaban artifact becomes available in artifactory, make sure to also send out an RB to update the internal Spark job type repo.

xkrogen replied:
We cannot. HadoopSecureSparkWrapper runs inside of a separate process which has had its classpath set properly by HadoopSparkJob. HadoopSparkJob, on the other hand, runs inside of the Azkaban process and shares the same classpath. The Azkaban main classpath does not include Spark JARs so I cannot access e.g. the SparkConf class.

Victsm replied:
That's reasonable. The changes lgtm.
